File: guides/micronaut-security-session-database-authentication/java/src/main/java/example/micronaut/UserFetcherService.java
package example.micronaut;

import example.micronaut.repositories.UserJdbcRepository;
import io.micronaut.core.annotation.NonNull;
import jakarta.inject.Singleton;
import jakarta.validation.constraints.NotBlank;

import java.util.Optional;

@Singleton // <1>
class UserFetcherService implements UserFetcher {

    private final UserJdbcRepository userJdbcRepository;

    UserFetcherService(UserJdbcRepository userJdbcRepository) { // <2>
        this.userJdbcRepository = userJdbcRepository;
    }

    @Override
    public Optional<UserState> findByUsername(@NotBlank @NonNull String username) {
        return userJdbcRepository.findByUsername(username).map(UserState.class::cast);
    }
}

File: guides/micronaut-database-authentication-provider/python/src/example/micronaut/user_fetcher_service.py
from java.util import Optional
from jakarta.inject import Singleton

from .user_fetcher import UserFetcher
from .user_jdbc_repository import UserJdbcRepository
from .user_state import UserState


@Singleton  # <1>
class UserFetcherService(UserFetcher):
    def __init__(self, user_jdbc_repository: UserJdbcRepository):  # <2>
        self.user_jdbc_repository = user_jdbc_repository

    def findByUsername(self, username: str) -> Optional[UserState]:
        return self.user_jdbc_repository.findByUsername(username)
